File: Trie/CodingNinjas/Moderate/CountDistinctSubstrings/CountDistinctSubstrings.cpp
// https://www.naukri.com/code360/problems/count-distinct-substrings_985292
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    Node *links[26];
    Node() {}

    bool containsKey(char ch)
    {
        return links[ch - 'a'] != nullptr;
    }

    void put(char ch, Node *node)
    {
        links[ch - 'a'] = node;
    }

    Node *get(char ch)
    {
        return links[ch - 'a'];
    }
};
class Trie
{
    Node *root;

public:
    Trie()
    {
        root = new Node();
    }

    int insertAndCount(string word)
    {
        Node *node = root;
        int count = 0;
        for (int i = 0; i < word.length(); i++)
        {
            if (!node->containsKey(word[i]))
            {
                count++;
                node->put(word[i], new Node());
            }
            node = node->get(word[i]);
        }
        return count;
    }
};
int countDistinctSubstrings(string &s)
{
    //    Write your code here.
    Trie *store = new Trie();
    int count = 0;
    for (int i = 0; i < s.length(); i++)
    {
        count += store->insertAndCount(s.substr(i));
    }
    return count + 1;
}

File: Trie/CodingNinjas/Moderate/CountDistinctSubstrings/CountDistinctSubstrings.java
// https://www.naukri.com/code360/problems/count-distinct-substrings_985292
class Trie {

    static class Node {

        Node links[];

        Node() {
            links = new Node[26];
        }

        boolean containsKey(char ch) {
            return links[ch - 'a'] != null;
        }

        Node get(char ch) {
            return links[ch - 'a'];
        }

        void put(char ch, Node node) {
            links[ch - 'a'] = node;
        }
    }

    Node root;

    Trie() {
        root = new Node();
    }

    int insertAndCount(String word) {
        Node node = root;
        int count = 0;
        for (int i = 0; i < word.length(); i++) {
            if (!node.containsKey(word.charAt(i))) {
                count++;
                node.put(word.charAt(i), new Node());
            }
            node = node.get(word.charAt(i));
        }
        return count;
    }
}

public class CountDistinctSubstrings {

    public static int countDistinctSubstrings(String s) {
        //	  Write your code here.
        Trie store = new Trie();
        int count = 1;
        for (int i = 0; i < s.length(); i++) {
            count += store.insertAndCount(s.substring(i));
        }
        return count;
    }
}
